### DISCUSSION TITLE
Finding elements in the opposite side of a square domain

### DISCUSSION THREAD

**Discussion by adhithya96** [libMesh/libmesh#3800] [General]
Is there any way to find the element on the  opposite side in libMesh ?

![SquareMesh](https://github.com/libMesh/libmesh/assets/89237852/56e0a1f9-af59-4289-a43f-f51d166ac0a8)

I was  able to find the nodes by using PointLocator but for elements I wasn't able to find anything on libMesh Doxygen. I am trying to use it to implement Bloch boundary condition in MOOSE.

**Comment by roystgnr** [comment]
libMesh optionally lets you construct a node-to-elem map; I believe MOOSE always constructs it.

**Reply by roystgnr** [reply]
Though, now that I think about it a little longer ... why *weren't* you able to find the elements by using PointLocator?  Searching for elements rather than nodes is the default behavior.

**Reply by adhithya96** [reply]
So I am am trying to use FunctorDirichletBC and FunctorNeumannBC to apply bloch boundary condition for displacement and stress.
```
  //Get point  locator  
  const auto pl = _mesh.getPointLocator();
  //Locate node on the other side of the geometry
  //The translation vector will be a user parameter
  libMesh::Point _translation_vec= libMesh::Point(_lattice_vec, 0.0, 0.0);
  const auto new_node = pl->locate_node(*_current_node + _translation_vec);

```
I am using this to find the opposite node for the dirichlet case but for NeumannBC I need to find the element. This is the code in FunctorNeumannBC.C which I am trying to modify'
```
  const Moose::ElemSideQpArg space_arg = {_current_elem, _current_side, _qp, _qrule, _q_point[_qp]};
  return _sign * _coef(space_arg, Moose::currentState()) *
         _functor(space_arg, Moose::currentState()) * _test[_i][_qp];
```
Here  instead of `_current_elem`, I need to find the _new_elem on the opposite side.

**Reply by roystgnr** [reply]
Check out the PointLocator docs; you can just use `operator()` in the simplest case.

**Reply by adhithya96** [reply]
While going through the PointLocator docs I found this ```virtual const Elem * 	operator() (const Point &p, const std::set< subdomain_id_type > *allowed_subdomains=nullptr) const =0```. When I try to modify the code I am getting the following error
```
/home/adhithyar/Documents/Project/static1d/src/BlochNeumannBCReal.C:55:3: error: 'virtual' outside class declaration
   55 |   virtual const Elem* _new_elem(*_current_node + _translation_vec);
      |   ^~~~~~~
/home/adhithyar/Documents/Project/static1d/src/BlochNeumannBCReal.C:55:34: error: '_current_node' was not declared in this scope; did you mean '_current_side'?
   55 |   virtual const Elem* _new_elem(*_current_node + _translation_vec);
      |                                  ^~~~~~~~~~~~~
      |                                  _current_side
/home/adhithyar/Documents/Project/static1d/src/BlochNeumannBCReal.C:61:53: error: '_new_side' was not declared in this scope
   61 |  const Moose::ElemSideQpArg space_arg = {_new_elem, _new_side, _qp, _qrule, _q_point[_qp]};
```
I tried adding the required headers and inheriting the base class which was used in the dirchlet case but I keep getting ```_current_node was declared in this scope```

**Reply by roystgnr** [reply]
We're digressing from "how do I use libMesh APIs" to "how do I use C++" at this point, I'm afraid.  For that I recommend reading Stroustrop cover-to-cover, though I admit I might have more time to be helpful myself if I didn't regularly do things like reading Stroustrop cover-to-cover...

You've possibly found the right API; you just need to call it.  Something like `const Elem * elem = (*pl)(*_current_node + _translation_vec);` would be the equivalent to the code you quoted originally.

On the other hand, I'm assuming that when you say "the element on the opposite side" you meant "*an* element on the opposite side".  The astute viewer of your diagram will note two elements on the opposite side of each node; in 3D you could have any number of elements sharing a vertex on the opposite side of a node.  If you're happy getting *any* of those then `operator()` really is what you want; if you want *all* of those we have other `PointLocator` APIs that will give you the whole batch.
